### PR TITLE
Feature/validate object

### PR DESCRIPTION
**Pull request by synapseradio**
_If there is a linked issue, mention it here._

- [ ] Bug
- [x] Feature

## Requirements

- [x] Wrote tests.
- [x] Updated docs and upgrade instructions, if necessary.

## Rationale

This PR adds a `validateObject` function to make it easier to validate larger structures like forms or API responses all in one go. Written with redux-form in mind.

Basically, you give it a matching schema to the one you're trying to work with, but with lists of rules instead of values. that returns a function over the form's values that will instantly give you the values, any errors, and a computed `hasErrors` property to make it easier to check for their existence.

This PR also fixes a couple of inconsistent messages and edge cases of strange formatting in some of the built in validators.

Check out docs and test for example use

### CODE REVIEW

File: src/__tests__/Validation.test.ts [hunk 1]
```diff
@@ -94,3 +101,48 @@ describe('Rule', () => {
...
 });
+
+describe('validateObject', () => {
+  it('can validate over an object, returning the original with an attached error object containing messages', () => {
+    const validate = validateObject({
+      // required fields
+      firstName: [isNonEmptyString],
+      lastName: [isNonEmptyString],
+      email: [isNonEmptyString, isValidEmail],
+      // non-required fields
```
Comment (willwhsiao) [inline src/__tests__/Validation.test.ts]: It's a small thing, but not sure if I get the semantics of the "non-required fields" here. Doesn't this still set hasErrors to `true` in the case where it's `undefined` or `null`?

Comment (synapseradio) [reply src/__tests__/Validation.test.ts]: that's a good question, it doesn't - but a form field would initialize to an empty string w or wo react

In a piece of arbitrary data you'd want an `isDefined` check
